spylib/multipass.py

```diff
@@ -0,0 +1,48 @@
+import datetime
+import json
+from base64 import urlsafe_b64encode
+from typing import Any, Dict
+
+from Crypto.Cipher import AES
+from Crypto.Hash import HMAC, SHA256
+from Crypto.Random import get_random_bytes
+from pydantic import BaseModel
+
+
+class CustomerData(BaseModel):
+    email: str
+
+
+def generate_token(secret: str, customer_data: Dict[str, Any]) -> bytes:
+    key = SHA256.new(secret.encode('utf-8')).digest()
+    encryption_key = key[0:16]
+    signature_key = key[16:32]
+
+    if 'email' not in customer_data:
+        raise ValueError('Missing email in customer data')
+
+    customer_data['created_at'] = datetime.datetime.utcnow().isoformat()
+    cypher_text = _encrypt(encryption_key, json.dumps(customer_data))
+    return urlsafe_b64encode(cypher_text + _sign(signature_key, cypher_text))
+
+
+def generate_url(secret: str, customer_data: Dict[str, Any], store_url) -> str:
+    token = generate_token(secret, customer_data).decode('utf-8')
+    return f'{store_url}/account/login/multipass/{token}'
+
+
+def _encrypt(encryption_key, plainText) -> bytes:
+    plainText = _pad(plainText)
+    iv = get_random_bytes(AES.block_size)
+    cipher = AES.new(encryption_key, AES.MODE_CBC, iv)
+    return iv + cipher.encrypt(str.encode(plainText))
+
+
+def _sign(signature_key, secret):
```

Comment:
I personally got confused by the `secret`, I thought it referred to the multipass secret. Maybe rename it to `cypher_text` to avoid the confusion?

Reply:
Maybe `secret` is not quite clear but it is what it's called in Shopify's documentation and UI:
![image](https://user-images.githubusercontent.com/6598959/169078040-303828d1-372d-4638-845f-6102871c495c.png)

Therefore we need to have this word. How about `multipass_secret` to make it very explicit?

Reply:
Oh I misread! My bad! You didn't change the `secret` I thought!  :wink: